src/coreclr/vm/method.cpp

```diff
@@ -1171,6 +1204,9 @@ COR_ILMETHOD* MethodDesc::GetILHeader()
     }
     CONTRACTL_END
 
+    if (IsRuntimeSupplied())
```

Comment:
Should this rather a be precondition to match `IsUnboxingStub`?

Reply:
I've made it a precondition.  One result of that is having to add a `HasILHeader` check in couple places that eagerly call `GetILHeader()`, but would be ok with getting `NULL`.

Reply:
I had to undo the precondition part. 

`HasILHeader() == false` does not prevent being asked for IL, because there are many ways how a method can provide IL, even if it has no RVA or header (i.e. IL can be dynamically generated or overriden via EnC or profiler patching), so we end up here for Async thunks. 

`IsUnboxingStub` has very special handling, so does not come here, but Async methods are treated as regular methods here, except the thunks do not provide static IL, which is then provided dynamically. 

The other changes that you mentioned are indeed unnecessary. 
As long as Async thunks report 0 in `GetRVA()`, that alone is sufficient.